crates/new_task/src/dependency.rs

```diff
@@ -0,0 +1,251 @@
+    /// Returns a vector of all of the results from it's dependencies.  
+    /// The results will be in a random order
+    async fn completed(self) -> Vec<T> {
+        let futures = FuturesUnordered::from_iter(self.deps);
+        futures.collect().await
+    }
+}
+
+impl<T: Clone + Send + Sync + 'static> AndDependency<T> {
+    /// Create from a vec of deps
+    #[must_use]
+    pub fn from_deps(deps: Vec<impl Dependency<T> + Send + 'static>) -> Self {
+        let mut pinned = vec![];
+        for dep in deps {
+            pinned.push(dep.completed().boxed());
+        }
+        Self { deps: pinned }
+    }
+    /// Add another dependency
+    pub fn add_dep(&mut self, dep: impl Dependency<T> + Send + 'static) {
+        self.deps.push(dep.completed().boxed());
+    }
+    /// Add multiple dependencies
+    pub fn add_deps(&mut self, deps: AndDependency<T>) {
+        for dep in deps.deps {
+            self.deps.push(dep);
+        }
+    }
+}
+
+/// Defines a dependency that complets when one of it's dependencies compeltes
+pub struct OrDependency<T> {
+    deps: Vec<BoxFuture<'static, T>>,
+}
+impl<T: Clone + Send + Sync> Dependency<T> for OrDependency<T> {
+    /// Returns the value of the first completed dependency
+    async fn completed(self) -> T {
+        let mut futures = FuturesUnordered::from_iter(self.deps);
+        loop {
+            if let Some(val) = futures.next().await {
+                break val;
+            }
+        }
+    }
+}
+
+impl<T: Clone + Send + Sync + 'static> OrDependency<T> {
+    /// Creat an `OrDependency` from a vec of dependencies
+    #[must_use]
+    pub fn from_deps(deps: Vec<impl Dependency<T> + Send + 'static>) -> Self {
+        let mut pinned = vec![];
+        for dep in deps {
+            pinned.push(dep.completed().boxed());
+        }
+        Self { deps: pinned }
+    }
+    /// Add another dependecy
+    pub fn add_dep(&mut self, dep: impl Dependency<T> + Send + 'static) {
+        self.deps.push(dep.completed().boxed());
+    }
+}
+
+/// A dependency that listens on a chanel for an event
+/// that matches what some value it wants.
+pub struct EventDependency<T: Clone + Send + Sync> {
+    /// Channel of incomming events
+    pub(crate) event_rx: Receiver<T>,
+    /// Closure which returns true if the incoming `T` is the
+    /// thing that completes this dependency
+    pub(crate) match_fn: Box<dyn Fn(&T) -> bool + Send>,
+}
+
+impl<T: Clone + Send + Sync + 'static> EventDependency<T> {
+    /// Create a new `EventDependency`
+    #[must_use]
+    pub fn new(receiver: Receiver<T>, match_fn: Box<dyn Fn(&T) -> bool + Send>) -> Self {
+        Self {
+            event_rx: receiver,
+            match_fn: Box::new(match_fn),
+        }
+    }
+}
+
+impl<T: Clone + Send + Sync + 'static> Dependency<T> for EventDependency<T> {
+    async fn completed(mut self) -> T {
+        loop {
+            let next = self.event_rx.recv().await.unwrap();
```

Comment:
This looks scary. Can we make this into `Dependency<Option<T>>` instead?

Reply:
Good Catch, yes this should be `Option<T>` or `Result<T>` and handle the errors appropriately